Урок 2. Анализ датасета с помощью Pandas
Задача 1: считать данные из файла

In [3]:
import pandas as pd

In [4]:
df=pd.read_csv('M:\GeekBrains\python\Data_Analityca\Jupiter\CSV\kc_house_data.csv', sep=',', encoding='UTF-8')

Вывести на экран первые 5 строк и посмотреть на описание признаков и их содержимое

In [5]:
df.head(5)


id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7      1180.0              0   
1      7242     2.0           0     0  ...      7      2170.0            400   
2     10000     1.0           0     0  ...      6       770.0              0   
3      5000     1.0           0     0  ...      7      1050.0            910   
4      8080     1.0           0     0  ...      8      1680.0              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

Задание 2. 
Проведите первичный анализ данных
Изучите типы данных, Найдите количество пропущенных ячеек в данных.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21611 non-null  float64
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

Посчитайте основные статистики по всем признакам и поизучайте их  
Пишите выводы

In [7]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21611.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.396095     291.509045   1971.005136     84.402258   
std        1.175459    828.128162     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

Выводы: данные представлены таблицей с 21613-ю записями по 20-ти столбцам. Все атрибуты заполнены на 100%, имеются данные трёх типов: object, целочисленные данные и числовые данные с плавающей запятой.  
В формате object предствалена Дата.  
Первичный анализ основных статистик показал, что такой параметр как цена, лучше привести к другим ед. измерения, чтобы отображаемые значения были читаемы.  
Количество спален в домах, представленных в таблице в среднем равно 3.37, 50% домов имеют не более трёх спален, однако изменяется их число в диапазоне от 0 до 33-ёх  
Количество ванных комнат изменяется в диапазоне от нуля до 8 и в среднем составляет 2.11, примечательно что некоторые квартиры обладают не целым числом ванных комнат, что вообще говоря не характерно для посчёта количества помещений.  
Жилая площадь домов изменяется в диапазоне от 290 единиц (квадратных метров?) и до 13540 единиц  
Каждый объект (дом) имеет оценку по отношению к близости к побережью, виду из окна, состоянию и т.д.

Задача 3.  
3.1 В каком диапазоне изменяются стоимости недвижимости?
3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?
3.3 Как много домов с разными этажами в данных?
3.4 Насколько хорошие состояния у домов в данных?
3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [12]:
df['price'].describe()

count    2.161300e+04
mean     5.400881e+05
std      3.671272e+05
min      7.500000e+04
25%      3.219500e+05
50%      4.500000e+05
75%      6.450000e+05
max      7.700000e+06
Name: price, dtype: float64

In [23]:
df['price'].astype('int64')

0        221900
1        538000
2        180000
3        604000
4        510000
          ...  
21608    360000
21609    400000
21610    402101
21611    400000
21612    325000
Name: price, Length: 21613, dtype: int64

In [28]:
df['price']=df['price'].astype('int64')

Изменим отображение числовых данных, чтобы они были более читаемы

In [27]:
pd.set_option('display.float_format', '{:.2f}'.format)

In [29]:
df['price'].describe()

count     21613.00
mean     540088.14
std      367127.20
min       75000.00
25%      321950.00
50%      450000.00
75%      645000.00
max     7700000.00
Name: price, dtype: float64

Cтоимость недвижимости изменяется  от 75000 условных единиц до 7700000 условных единиц

3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

In [30]:
df['sqft_living'].describe()

count   21613.00
mean     2079.90
std       918.44
min       290.00
25%      1427.00
50%      1910.00
75%      2550.00
max     13540.00
Name: sqft_living, dtype: float64

In [33]:
df['sqft_lot'].describe()

count     21613.00
mean      15106.97
std       41420.51
min         520.00
25%        5040.00
50%        7618.00
75%       10688.00
max     1651359.00
Name: sqft_lot, dtype: float64

In [44]:
df['living_area_ratio'] = df['sqft_living']/df['sqft_lot']


In [45]:
df[['sqft_living','sqft_lot','living_area_ratio']].head()

sqft_living  sqft_lot  living_area_ratio
0         1180      5650               0.21
1         2570      7242               0.35
2          770     10000               0.08
3         1960      5000               0.39
4         1680      8080               0.21

In [46]:
df['living_area_ratio'].describe()

count   21613.00
mean        0.32
std         0.27
min         0.00
25%         0.16
50%         0.25
75%         0.41
max         4.65
Name: living_area_ratio, dtype: float64

In [ ]:
В среднем жилая площадь занимает 0,32% от общей по всем домам, однако существуют дома, где жилая площадь больше чем общая, что мне кажется ошибочным явлением

In [49]:
df[df['sqft_living']>df['sqft_lot']]


id             date   price  bedrooms  bathrooms  sqft_living  \
63     9528102996  20141207T000000  549000         3       1.75         1540   
116    3362400431  20140626T000000  518500         3       3.50         1590   
175    7234600798  20150210T000000  425000         3       2.50         1120   
406    7821200390  20140806T000000  450000         3       2.00         1290   
547    7899800860  20150319T000000  259950         2       2.00         1070   
...           ...              ...     ...       ...        ...          ...   
21595  1972201967  20141031T000000  520000         2       2.25         1530   
21601  5100403806  20150407T000000  467000         3       2.50         1425   
21604  9834201367  20150126T000000  429000         3       2.00         1490   
21607  2997800021  20150219T000000  475000         3       2.50         1310   
21608   263000018  20140521T000000  360000         3       2.50         1530   

       sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
63         1044    3.00           0     0  ...     1540.00              0   
116        1102    3.00           0     0  ...     1590.00              0   
175        1100    2.00           0     0  ...      820.00            300   
406        1213    3.00           0     0  ...     1290.00              0   
547         649    2.00           0     0  ...      720.00            350   
...         ...     ...         ...   ...  ...         ...            ...   
21595       981    3.00           0     0  ...     1480.00             50   
21601      1179    3.00           0     0  ...     1425.00              0   
21604      1126    3.00           0     0  ...     1490.00              0   
21607      1294    2.00           0     0  ...     1180.00            130   
21608      1131    3.00           0     0  ...     1530.00              0   

       yr_built  yr_renovated  zipcode   lat    long  sqft_living15  \
63         2014             0    98115 47.68 -122.32           1580   
116        2010             0    98103 47.68 -122.35           1620   
175        2008             0    98122 47.61 -122.31           1590   
406        2001             0    98103 47.66 -122.34           1290   
547        2008             0    98106 47.52 -122.36           1070   
...         ...           ...      ...   ...     ...            ...   
21595      2006             0    98103 47.65 -122.35           1530   
21601      2008             0    98125 47.70 -122.32           1285   
21604      2014             0    98144 47.57 -122.29           1400   
21607      2008             0    98116 47.58 -122.41           1330   
21608      2009             0    98103 47.70 -122.35           1530   

       sqft_lot15  living_area_ratio  
63           3090               1.48  
116          3166               1.44  
175          1795               1.02  
406          3235               1.06  
547           928               1.65  
...           ...                ...  
21595        1282               1.56  
21601        1253               1.21  
21604        1230               1.32  
21607        1265               1.01  
21608        1509               1.35  

[789 rows x 22 columns]

Количество домов, где жилая площадь больше общей площади 789 объектов, либо я не правильно опнимаю понятие жилой площади или общей площади, либо в данных есть логические ошибки.

3.3 Как много домов с разными этажами в данных?

In [50]:
df['floors'].value_counts()

1.00    10680
2.00     8241
1.50     1910
3.00      613
2.50      161
3.50        8
Name: floors, dtype: int64

опять таки не понятно что значит 'пол этажа'

3.4 Насколько хорошие состояния у домов в данных?

In [51]:
df['condition'].value_counts()

3    14031
4     5679
5     1701
2      172
1       30
Name: condition, dtype: int64

In [ ]:
3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [52]:
df['yr_built'].min()

1900

In [53]:
df['yr_built'].max()

2015

Ответьте на несколько вопросов

4.1 Сколько в среднем стоят дома, у которых 2 спальни?
4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?
4.3 Как много домов коснулся ремонт?
4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?


4.1 Сколько в среднем стоят дома, у которых 2 спальни?

In [57]:
df[df['bedrooms']==2]['price'].mean()

401372.681884058

4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [59]:
df[df['price']>600000]['sqft_lot'].mean()

20442.524776214832

4.3 Как много домов коснулся ремонт?

In [66]:
df[df['yr_renovated']>0]['id'].count()

914

In [ ]:
4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [70]:
first_price = df[df['grade']>10]['price'].mean()
first_price

1678635.1175298805

In [73]:
sec_price = df[df['grade']<4]['price'].mean()
sec_price

189750.0

In [74]:
first_price - sec_price

1488885.1175298805

Ответьте на несколько вопросов
5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?
5.2 Выберите дом клиенту
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?
5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?

In [78]:
df[(df['waterfront']==1) & (df['bathrooms']>2) & (df['sqft_basement']>0)]['id'].count()

76

5.2 Выберите дом клиенту
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?

In [79]:
df['view'].describe()

count   21613.00
mean        0.23
std         0.77
min         0.00
25%         0.00
50%         0.00
75%         0.00
max         4.00
Name: view, dtype: float64

In [84]:
df['view'].value_counts()

0    19489
2      963
3      510
1      332
4      319
Name: view, dtype: int64

Так как понятие "очень красивый вид" не формализовано, создадим переменную.  
Пускай будет градация: 0 - неудовлетворительно, 1 - нормально, 2 красиво, 3 - очень красиво, 4 - превосходно

In [89]:
beautiful_formalization = 3

Если у кого-то будет другое мнение и шкала, можно поменять переменную

In [86]:
df['condition'].describe()

count   21613.00
mean        3.41
std         0.65
min         1.00
25%         3.00
50%         3.00
75%         4.00
max         5.00
Name: condition, dtype: float64

так же создадим переменную для "очень хорошего состояния"

In [90]:
very_good_condition = 4

In [105]:
df[((df['view']>=beautiful_formalization) | (df['waterfront']==1)) & (df['condition']>=very_good_condition)& (df['yr_built']>=1980)]['price'].describe()

count        55.00
mean    1637253.42
std     1057605.44
min      420000.00
25%      829500.00
50%     1411600.00
75%     2160500.00
max     5300000.00
Name: price, dtype: float64

min      420000.00  
max     5300000.00


5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [110]:
df[(df['sqft_basement']==0) & (df['floors']==2) & (df['price']<=150000)]['condition'].mean()

3.0